### PR TITLE
DXF: add creation option to set  and  DXF system variables

### LINKED ISSUE #11423: DXF Driver - Allow Setting of Units in Header
### Feature description

Currently, when using the DXF Writer driver, the dxf header comes from a static file. This means that the unit of measurement is set to a static value, and must be modified manually for each use of the DXF writer, either before its use by modifying the header template, or after by modifying the generated DXF file, as is clearly stated in your documentation. 

I would like to instead be able to set the $INSUNIT and $MEASUREMENT values programmatically as a part of the DXF Writing process, so that manual editing of these values is not required.

### Additional context

The ideal outcome for this would be to allow an ogr2ogr command that is converting a vector layer to DXF to export the correct $INSUNITS and $MEASUREMENT unit in the header that matches the units of the CRS of the output. 

Implementing this would also allow QGIS vector layer exporting to work as expected and not default to inches no matter the destination CRS. 

I spent hours trying to figure out why the DXF export in QGIS was not lining up with the rest of my drawing until I found the gis stack exchange question on the topic from over 8 years ago explaining this limitation. The QGIS team has apparently marked this bug as an upstream issue with gdal and does not appear to be willing to find an alternative solution.

### PR DESCRIPTION
**Pull request by rouault**
Fixes #11423

### CODE REVIEW

**Comment by rouault** [comment]
@atlight I was wondering if using the CRS unit as a default for INSUNITS would be a sane default ? And what about MEASUREMENT (should it be related to CRS units?) ? For now, I've kept the values from header.dxf (inches / imperial) as the default

File: ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp [hunk 1]
```diff
@@ -529,6 +539,151 @@ bool OGRDXFWriterDS::TransferUpdateHeader(VSILFILE *fpOut)
...
+                    {
+                        const char *pszValue;
+                        int nValue;
+                    } INSUNITSMap[] = {
+                        {"UNITLESS", 0},
+                        {"INCHES", 1},
+                        {"FEET", 2},
+                        {"MILLIMETERS", 4},
+                        {"CENTIMETERS", 5},
+                        {"METERS", 6},
```
Comment (elpaso) [inline ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp:608]: Maybe we should be extra tolerant and also accept the UK spelling METRES.

**Review by elpaso — APPROVED** [review]

**Review by atlight — COMMENTED** [review]
Haven't tested, but looks good to me. Just minor stuff here.

**Comment by atlight** [comment]
> @atlight I was wondering if using the CRS unit as a default for INSUNITS would be a sane default ? And what about MEASUREMENT (should it be related to CRS units?) ? For now, I've kept the values from header.dxf (inches / imperial) as the default

(Reading your comment after leaving my review) I think this makes a lot of sense. If your CRS is in metres you ideally should get a DXF in metres without taking any active steps to ensure that.

Not really sure about MEASUREMENT, it doesn't seem very important. It's not something my company or our clients ever use.

File: ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp [hunk 2]
```diff
@@ -529,6 +539,151 @@ bool OGRDXFWriterDS::TransferUpdateHeader(VSILFILE *fpOut)
             }
         }
 
+        // Patch INSUNITS
+        if (nCode == 9 && EQUAL(szLineBuf, "$INSUNITS") &&
```
[the remark this answers is not in the document]
Comment (rouault) [reply ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp:543]: I also agree that when actually debugging code and putting side to side code, files and DXF documentation, having those literal numeric codes makes thing actually easier than using named constants.

File: ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp (hunk 1 above)
Comment (rouault) [reply ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp:592]: The declaration of the option uses an enumeration, so if users misspell they'll get an error. I'm not too entusiastic about tolerating all possible spelling variations (UK vs US, plural vs singular). The ones here are the ones of the DXF documentation

File: ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp [hunk 3]
```diff
@@ -529,6 +539,151 @@ bool OGRDXFWriterDS::TransferUpdateHeader(VSILFILE *fpOut)
...
+                    {
+                        CPLError(CE_Warning, CPLE_AppDefined,
+                                 "Layer has no CRS. Using default value of "
+                                 "INSUNIT from template header file");
+                    }
+                    else if (m_oSRS.IsProjected())
+                    {
+                        const char *pszUnits = nullptr;
+                        const double dfUnits = m_oSRS.GetLinearUnits(&pszUnits);
+                        if (std::fabs(dfUnits - 1) <= 1e-10)
```
[the remark this answers is not in the document]
Comment (rouault) [reply ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp]: yes, I've used a lamba IsAlmostEqual() to make it clearer

File: ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp [hunk 4]
```diff
@@ -529,6 +539,151 @@ bool OGRDXFWriterDS::TransferUpdateHeader(VSILFILE *fpOut)
...
+            if (nCode == 70)
+            {
+                int nVal = -1;
+                if (m_osINSUNITS == "FROM_CRS")
+                {
+                    if (m_oSRS.IsEmpty())
+                    {
+                        CPLError(CE_Warning, CPLE_AppDefined,
+                                 "Layer has no CRS. Using default value of "
+                                 "INSUNIT from template header file");
```
[the remark this answers is not in the document]
Comment (rouault) [reply ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp]: fixed

File: ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp [hunk 5]
```diff
@@ -529,6 +539,151 @@ bool OGRDXFWriterDS::TransferUpdateHeader(VSILFILE *fpOut)
...
+                                           CPLAtof(SRS_UL_US_FOOT_CONV)) <=
+                                 1e-10)
+                        {
+                            nVal = 21;
+                        }
+                        else
+                        {
+                            CPLError(CE_Warning, CPLE_AppDefined,
+                                     "Could not translate CRS unit %s to "
+                                     "INSUNIT. Using default value from "
```
[the remark this answers is not in the document]
Comment (rouault) [reply ogr/ogrsf_frmts/dxf/ogrdxfwriterds.cpp]: fixed

**Comment by rouault** [comment]
> I think this makes a lot of sense. If your CRS is in metres you ideally should get a DXF in metres without taking any active steps to ensure that.

ok, I've remove the INSUNIT=FROM_CRS mode, and rather use a default INSUNIT=AUTO mode that tries to use the CRS info when available, and fallbacks to the value of the template header.dxf file

**Comment by agiudiceandrea** [comment]
See https://github.com/qgis/QGIS/pull/62339.

**Pull request merged**